### PR TITLE
fix(coap): Add config option to enable/disable Q-Block option

### PR DESCRIPTION
**Pull request by hmalpani**
Add config option to enable/disable Q-Block option

Fixes: https://github.com/espressif/idf-extra-components/issues/324

### CODE REVIEW

**Comment by mahavirj** [comment]
@mrdeep1 

- Can you please take a look at this PR?
- When can we expect new bugfix release of libcoap containing the fix for `CVE-2024-0962`?

File: coap/Kconfig [hunk 1]
```diff
@@ -102,6 +102,12 @@ menu "CoAP Configuration"
 
             If this option is disabled, redundant CoAP Observe Persist code is removed.
 
+    config COAP_Q_BLOCK
+        bool "Enable block-wise transfer in CoAP"
```
Comment (mrdeep1) [inline coap/Kconfig:106]: bool "Enable Q-Block (RFC9177) support in CoAP"

Note: change as regular block transfers (RFC7959) is always enabled

File: coap/Kconfig [hunk 2]
```diff
@@ -102,6 +102,12 @@ menu "CoAP Configuration"
...
+        default n
+        help
+            Enable Block-Wise Transfers support for CoAP
```
Comment (mrdeep1) [inline coap/Kconfig:109]: Enable Q-Block (RFC9177) fast block transfers using NON (instead of CON) support for CoAP

**Review by mrdeep1 — CHANGES_REQUESTED** [review]
Q-Block supports fast block based transfer / recovery when using NON PDU type (instead of CON) (RFC9177),  This is in addition to block transfer CON support in RFC7959 which is always enabled in CoAP.

**Comment by mrdeep1** [comment]
> When can we expect new bugfix release of libcoap containing the fix for CVE-2024-0962?

There is a tagged libcoap release that has this fix in - which I can use to update idf-extra-components.  In the near term, we are planning on getting out a Release Candidate which covers this CVE with other updates as well which I will be putting into idf-extra-components.

Do you want me to update the code with the tagged fix?

**Comment by mahavirj** [comment]
> > When can we expect new bugfix release of libcoap containing the fix for CVE-2024-0962?
> 
> There is a tagged libcoap release that has this fix in - which I can use to update idf-extra-components. In the near term, we are planning on getting out a Release Candidate which covers this CVE with other updates as well which I will be putting into idf-extra-components.
> 
> Do you want me to update the code with the tagged fix?

I thought upgrade to official `4.3.5` would be a cleaner solution. It will also help to align our component version.

At the moment, I see only `4.3.4a` tag here https://github.com/obgm/libcoap/tags. If we plan to upgrade to this tag, then I am not sure what will be the corresponding IDF component version.

**Comment by mrdeep1** [comment]
Yes, `4.3.4a` contains the CVE fix.  I had assumed that `coap/idf_component.yml` version would become "4.3.4~2", but I guess something else might not like `4.3.4a`.

Certainly` 4.3.5` would be cleaner.

File: coap/Kconfig (hunk 1 above)
Comment (mrdeep1) [reply coap/Kconfig]: Change is good

File: coap/Kconfig (hunk 2 above)
Comment (mrdeep1) [reply coap/Kconfig]: Change is good

**Review by mahavirj — APPROVED** [review]

**Pull request merged**